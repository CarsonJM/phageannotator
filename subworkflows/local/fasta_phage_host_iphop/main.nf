//
// Predict phage hosts using iPHoP
//

include { IPHOP_DOWNLOAD    } from '../../../modules/nf-core/iphop/download/main'
include { UNTAR             } from '../../../modules/nf-core/untar/main'
include { UNTAR as UNTAR2   } from '../../../modules/nf-core/untar/main'
include { GUNZIP            } from '../../../modules/nf-core/gunzip/main'
include { IPHOP_PREDICT     } from '../../../modules/nf-core/iphop/predict/main'

workflow FASTA_PHAGE_HOST_IPHOP {
    take:
    virus_fasta_gz  // [ [ meta ], fasta.gz ]   , virus sequences (mandatory)
    iphop_db        // [ checkv_db ]            , iPHoP database directory (optional)

    main:
    ch_versions = Channel.empty()

    // if iphop_db exists, skip IPHOP_DOWNLOAD
    if ( iphop_db ){
        ch_iphop_db = iphop_db
    } else {
        // if partial run is specified the minimal, incomplete database will be downloaded
        if ( !params.iphop_partial_test ) {
            //
            // MODULE: download standard IPHoP database
            //
            ch_iphop_db = IPHOP_DOWNLOAD( ).iphop_db
            ch_versions = ch_versions.mix(IPHOP_DOWNLOAD.out.versions)
        } else {
            //
            // MODULE: unpack minimal iphop db
            //
            ch_iphop_db = UNTAR ( [ [ id:'iphop_minimal_db' ], file(params.test_data['modules_nfcore']['iphop_test_db_tar_gz'], checkIfExists: true) ] ).untar.map { it[1] }
            ch_versions = ch_versions.mix(UNTAR.out.versions)
        }
    }

    //
    // MODULE: Gunzip fasta for IPHoP
    //
    ch_viruses_fasta = GUNZIP ( virus_fasta_gz ).gunzip
    ch_versions = ch_versions.mix(GUNZIP.out.versions.first())

    // create input for partial iphop run
    ch_iphop_partial_input = Channel.empty()
    if ( params.iphop_partial_test ) {
        //
        // MODULE: unpack data for partial iphop run
        //
        ch_iphop_partial_input = UNTAR2 ( [ [ id:'iphop_partial_input' ], file(params.test_data['modules_nfcore']['iphop_data_tar_gz'], checkIfExists: true) ] ).untar.map { it[1] }
        ch_versions = ch_versions.mix(UNTAR2.out.versions)
    }

    //
    // MODULE: Predict virus host
    //
    IPHOP_PREDICT ( ch_viruses_fasta, ch_iphop_db, ch_iphop_partial_input )
    ch_host_predictions_tsv  = IPHOP_PREDICT.out.iphop_genus
    ch_versions = ch_versions.mix(IPHOP_PREDICT.out.versions.first())

    emit:
    host_predictions_tsv    = ch_host_predictions_tsv  // [ [ meta ], genus_predictions.tsv ]   , TSV file containing host genus predictions
    versions                = ch_versions               // [ versions.yml ]
}
