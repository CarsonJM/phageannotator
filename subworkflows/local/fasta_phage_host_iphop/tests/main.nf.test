nextflow_workflow {

    name "Test Subworkflow: FASTA_PHAGE_HOST_IPHOP"
    script "subworkflows/local/fasta_phage_host_iphop/main.nf"
    workflow "FASTA_PHAGE_HOST_IPHOP"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fasta_phage_host_iphop"
    tag "fasta_phage_host_iphop_default"


    test("Parameters: default") {

        when {
            params {
                outdir              = "$outputDir"
                iphop_test_db       = true
            }
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], [ file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists: true) ] ])
                input[1] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
