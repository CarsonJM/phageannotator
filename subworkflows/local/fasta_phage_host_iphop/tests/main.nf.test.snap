{
    "Parameters: default": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "Host_prediction_to_genus_m90.csv:md5,e529062e32656f9e9e1737ebd66c4e83"
                    ]
                ],
                "1": [
                    "versions.yml:md5,41c2135a3578a7a52a23fd3243c22e80",
                    "versions.yml:md5,9a3afe922371942954848605293c0700",
                    "versions.yml:md5,bded68aef8ec18c06204c83910c8a760"
                ],
                "host_predictions_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "Host_prediction_to_genus_m90.csv:md5,e529062e32656f9e9e1737ebd66c4e83"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,41c2135a3578a7a52a23fd3243c22e80",
                    "versions.yml:md5,9a3afe922371942954848605293c0700",
                    "versions.yml:md5,bded68aef8ec18c06204c83910c8a760"
                ]
            }
        ],
        "timestamp": "2023-12-05T12:39:29.286383941"
    }
}